### PR TITLE
Add support for embedded Prometheus endpoint in Keeper

### LINKED ISSUE #34548: Embedded Prometheus endpoint for clickhouse-keeper
**Use case**

Currently, clickhouse-keeper could be monitored only via `echo "mntr" | nc localhost 2181`, and provide not so much metrics as i expected.
For example, we need to know about raft errors.

**Describe the solution you'd like**

I propose implements embedded prometheus endpoint for standalone clickhouse-keeper to allow us monitor clickhouse-keeper properly

### PR DESCRIPTION
**Pull request by antonio2368**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
 Keeper feature: add support for embedded Prometheus endpoint.

Close https://github.com/ClickHouse/ClickHouse/issues/34548

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: programs/keeper/Keeper.cpp [hunk 1]
```diff
@@ -52,6 +58,14 @@ int mainEntryClickHouseKeeper(int argc, char ** argv)
     }
 }
 
+#ifdef KEEPER_STANDALONE_BUILD
+
+void collectCrashLog(
+    Int32, UInt64, const String &, const StackTrace &)
+{}
```
Comment (antonio2368) [inline programs/keeper/Keeper.cpp:65]: It seems Darwin doesn't like weak symbols, the standalone build failed on my Mac.

Comment (serxa) [reply programs/keeper/Keeper.cpp:65]: Maybe leave a comment in the code about it?

File: programs/keeper/Keeper.cpp [hunk 2]
```diff
@@ -261,6 +275,60 @@ void Keeper::defineOptions(Poco::Util::OptionSet & options)
...
+    {
+        return context.getConfigRef().getUInt64("keeper_server.http_receive_timeout", DEFAULT_HTTP_READ_BUFFER_TIMEOUT);
+    }
+
+    Poco::Timespan getSendTimeout() const override
+    {
+        return context.getConfigRef().getUInt64("keeper_server.http_send_timeout", DEFAULT_HTTP_READ_BUFFER_TIMEOUT);
+    }
+
+    const TinyContext & context;
```
Comment (serxa) [inline programs/keeper/Keeper.cpp:324]: Is it okay to have it here just by reference? looks a little bit dangerous to me... but I'm not sure about TinyContext ownership logic in Keeper

File: programs/keeper/Keeper.cpp [hunk 3]
```diff
@@ -401,6 +485,29 @@ try
...
+        http_params->setTimeout(DBMS_DEFAULT_RECEIVE_TIMEOUT_SEC);
+        http_params->setKeepAliveTimeout(keep_alive_timeout);
+
+        /// Prometheus (if defined and not setup yet with http_port)
+        port_name = "prometheus.port";
+        createServer(listen_host, port_name, listen_try, [&](UInt16 port)
+        {
+            Poco::Net::ServerSocket socket;
+            auto address = socketBindListen(socket, listen_host, port);
+            // TODO(antonio2368): use config
```
Comment (serxa) [inline programs/keeper/Keeper.cpp:501]: left for the next round?

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 4]
```diff
@@ -1167,6 +1167,7 @@ BlockIO InterpreterCreateQuery::createTable(ASTCreateQuery & create)
             getContext()
         ))
         {
+            LOG_INFO(&Poco::Logger::get("LOGGER"), "{}", create.select->dumpTree());
```
Comment (serxa) [inline src/Interpreters/InterpreterCreateQuery.cpp:1170]: looks unrelated

**Review by serxa — COMMENTED** [review]
Would be good to have the simplest possible integration test for this feature if it is not too complex

**Comment by antonio2368** [comment]
> Would be good to have the simplest possible integration test for this feature if it is not too complex

I'm not sure how to start a standalone Keeper to test out the feature. Integration tests start CH servers with embedded Keeper. 😕

File: programs/keeper/Keeper.cpp (hunk 3 above)
Comment (antonio2368) [reply programs/keeper/Keeper.cpp]: No, just forgot, thanks!

File: programs/keeper/Keeper.cpp (hunk 2 above)
Comment (antonio2368) [reply programs/keeper/Keeper.cpp]: TinyContextPtr is better.

**Review by serxa — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
So, if I have an embedded Keeper, is there any way to access this handler?

**Comment by antonio2368** [comment]
> So, if I have an embedded Keeper, is there any way to access this handler?

The information provided through Prometheus for standalone Keeper is only subset of the information provided from server. So if you have embedded Keeeper, you get the exact information using server's Prometheus endpoint.
